In [237]:
%env KERAS_BACKEND=tensorflow

env: KERAS_BACKEND=tensorflow


In [238]:
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt

In [239]:
#讀入資料庫
from keras.datasets import mnist

In [240]:
(x_train, y_train),(x_test, y_test) = mnist.load_data()

In [241]:
x_train[1234].shape # 28 * 28

(28, 28)

In [242]:
# CNN要(28, 28, 1)
x_train = x_train.reshape(60000, 28, 28, 1)
x_test = x_test.reshape(10000, 28, 28, 1)

In [243]:
x_train[1234].shape

(28, 28, 1)

In [244]:
from keras.utils import np_utils

In [245]:
y_train = np_utils.to_categorical(y_train, 10)
y_test = np_utils.to_categorical(y_test, 10)

In [246]:
x_train = x_train / 255
x_test = x_test/255

In [247]:
from keras.models import Sequential
from keras.layers import Dense, Activation, Flatten
from keras.layers import Conv2D, MaxPool2D
from keras.optimizers import SGD

In [248]:
# 建構神經網路
model = Sequential()

In [249]:
model.add(Conv2D(8,(10,10),padding='same',input_shape=(28,28,1)))


In [250]:
model.add(Activation('relu'))

In [251]:
model.add(MaxPool2D(pool_size=(3,3)))

In [252]:
model.add(Conv2D(16,(10,10),padding='same'))
model.add(Activation('relu'))
model.add(MaxPool2D(pool_size=(3,3)))
model.add(Conv2D(32,(10,10),padding='same'))
model.add(Activation('relu'))
model.add(MaxPool2D(pool_size=(3,3)))

In [253]:
model.add(Flatten())
model.add(Dense(15))
model.add(Activation('relu'))
model.add(Dense(300))
model.add(Activation('relu'))
model.add(Dense(10))
model.add(Activation('softmax'))

In [254]:
model.compile(loss="categorical_crossentropy",
              optimizer='Adam',
              metrics=['accuracy'])

In [255]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_28 (Conv2D)           (None, 28, 28, 8)         808       
_________________________________________________________________
activation_49 (Activation)   (None, 28, 28, 8)         0         
_________________________________________________________________
max_pooling2d_28 (MaxPooling (None, 9, 9, 8)           0         
_________________________________________________________________
conv2d_29 (Conv2D)           (None, 9, 9, 16)          12816     
_________________________________________________________________
activation_50 (Activation)   (None, 9, 9, 16)          0         
_________________________________________________________________
max_pooling2d_29 (MaxPooling (None, 3, 3, 16)          0         
_________________________________________________________________
conv2d_30 (Conv2D)           (None, 3, 3, 32)          51232     
__________

In [256]:
model.fit(x_train, y_train, batch_size=1000, epochs=10)

Epoch 1/10
60000/60000 [==============================] - 35s 591us/step - loss: 1.4701 - acc: 0.4656
Epoch 2/10
60000/60000 [==============================] - 35s 590us/step - loss: 0.4925 - acc: 0.8482
Epoch 3/10
60000/60000 [==============================] - 38s 627us/step - loss: 0.2852 - acc: 0.9134
Epoch 4/10
60000/60000 [==============================] - 38s 627us/step - loss: 0.2017 - acc: 0.9406
Epoch 5/10
60000/60000 [==============================] - 36s 605us/step - loss: 0.1572 - acc: 0.9530
Epoch 6/10
60000/60000 [==============================] - 38s 637us/step - loss: 0.1304 - acc: 0.9607
Epoch 7/10
60000/60000 [==============================] - 37s 624us/step - loss: 0.1167 - acc: 0.9641
Epoch 8/10
60000/60000 [==============================] - 37s 610us/step - loss: 0.1026 - acc: 0.9689
Epoch 9/10
60000/60000 [==============================] - 38s 635us/step - loss: 0.0925 - acc: 0.9720
Epoch 10/10
60000/60000 [==============================] - 37s 624us/step - loss: 

In [257]:
score = model.evaluate(x_test, y_test)

10000/10000 [==============================] - 3s 323us/step


In [258]:
print(f'測試資料的 loss: {score[0]:.5f}')
print(f'測試資料的正確率: {score[1]}')

測試資料的 loss: 0.08012
測試資料的正確率: 0.9736


## 結論
老師的值：

Conv2D(4, (5, 5)) (8,(5,5)) (32,(5,5)) Dense 9 -> 10 acc:0.86

我試了很多次不同的數值：

optimizer為原本的

Conv2D(8,(6,6) (16,(6,6) (32,(6,6) Dense 40->60->10 acc: 0.1124

Conv2D(8, (10, 10)) (16,(10,10)) (32,(10,10)) Dense 9 -> 10 acc:0.96 備註：跑的速度極慢

最意外的是最後一組

Conv2D(8, (10, 10)) (16,(10,10)) (32,(10,10)) Dense 15 -> 300 -> 10

optimizer為Adam時，acc:0.9736

optimizer為原本的時候，acc:0.12

因此改變optimizer的效益會最大。

